### PR TITLE
General cleanup

### PR DESCRIPTION
**Pull request by mahesh-hegde**
A pretty large cleanup PR. Updates some code that's untidy, obtuse or outright criminal.

_jni:_
* Update `jni:setup` script to not rebuild already built libraries, and remove unused options.
* Update READMEs with better instructions about setup.
* Rename several variables, functions and files.
* Use `JniResult` in reflective API, this probably belonged to separate PR sorry 😞 
* remove unused / not-usable variables from generated ffigen bindings to C JNI API.
* Move `global_jni_env` to `src/third_party` and add Android licenses.
* Better comment the C parts of jni library
* Update patched ffigen to not generate __va_list_tag or whatever.

_jnigen:_
* Delete API based config in favor of YAML config from PDFBox example.
* Move more logic into common.dart, so that it will be reusable with pure dart bindings in a future PR.
* Move `PackagePathResolver` to `FilesWriter` implementation, in separate file, because these 2 are kinda coupled.
* Use try ... finally for deleting files in `android_sdk_tools.dart`. Not error on SDK_ROOT not set, if gradle option is given.
* use `-q` option with maven and gradle (for less output).
* Refactor BindingsWriter into abstract class in separate file.
* Add reason parameter for tests comparing config struct.
* Add `generateAndCompareBindings` and `generateAndAnalyzeBindings` primitives to `test_util.dart`, reducing boilerplate in individual tests.

### CODE REVIEW

**Comment by mahesh-hegde** [comment]
1 test failing due to linux and windows header file differences.

**Comment by dcharkes** [comment]
RE: PR description

> use `-q` option with maven and gradle (for less output).

Alternatively, we could chose to _not_ stream the output of those tools to the standard out, but buffer it in Dart, and only put it on standard error _if_ we get an exception.

Also, if we have a verbosity command-line option, we could decide to use `-q` or not based on that option.

File: README.md [hunk 1]
```diff
@@ -9,16 +9,19 @@ This project intends to provide 2 packages to enable JNI interop from Dart & Flu
 | [jni](jni/) | Ergonomic C bindings to JNI C API and several helper methods |
 | [jnigen](jnigen/) | Tool to generate Dart bindings to Java code using FFI |
 
-This is a work-in-progress project under Google Summer of Code 2022 program.
+This project is a work-in-progress and highly experimental. Users can check out and experiment with the examples. We do not guarantee stability of the API or commands at this point.
 
 ## SDK Requirements
 Dart standalone target is supported, but due to some problems with pubspec, the `dart` command must be from Flutter SDK and not Dart SDK. See [dart-lang/pub#3563](https://github.com/dart-lang/pub/issues/3563).
 
+Along with JDK, maven (`mvn` command) is required. On Windows, it can be installed using a package manager such as `scoop`.
```
Comment (dcharkes) [inline README.md:17]: In the Dart ecosystem we recommend Chocolatey (see https://dart.dev/get-dart#install -> Windows). So maybe refer to that instead.

File: jni/lib/src/jni_object.dart [hunk 2]
```diff
@@ -162,14 +97,14 @@ int _getCallType(int? callType, int defaultType, Set<int> allowed) {
   throw InvalidCallTypeException(callType, allowed);
 }
 
-T _callOrGet<T>(int? callType, Function(int) f) {
+T _callOrGet<T>(int? callType, JniResult Function(int) f) {
```
Comment (dcharkes) [inline jni/lib/src/jni_object.dart:100]: Maybe give `f` a better name, then the calls with `f(...)` are a bit easier to see further down below.

File: jni/lib/src/jni_object.dart [hunk 3]
```diff
@@ -203,33 +160,27 @@ T _callOrGet<T>(int? callType, Function(int) f) {
...
     default:
       throw UnsupportedError('Unknown type $T');
   }
   return result;
 }
 
-T _callMethod<T>(
-        int? callType, List<dynamic> args, Function(int, Pointer<JValue>) f) =>
+T _callMethod<T>(int? callType, List<dynamic> args,
+        JniResult Function(int, Pointer<JValue>) f) =>
```
Comment (dcharkes) [inline jni/lib/src/jni_object.dart:175]: Same: `f`.

File: jni/src/README.md [hunk 4]
```diff
@@ -1,27 +1,5 @@
...
-Run `cmake --build` command from your source, prefixed with `bear`.
-
-`bear -- cmake --build <build_dir>`
-
-On some distro versions of `bear` command, the `--` needs to be omitted.
-
-`bear cmake --build <build_dir>`
+## LSP instructions
 
+see https://releases.llvm.org/8.0.0/tools/clang/tools/extra/docs/clangd/Installation.html#project-setup for creating compile_commands.json for use with LSP based editor plugins (in VS Code, VIM etc..).
```
Comment (dcharkes) [inline jni/src/README.md:5]: https://github.com/dart-lang/jnigen/issues/81

File: jnigen/README.md [hunk 5]
```diff
@@ -14,6 +14,8 @@ Basic features of the Java language (static and instance methods, fields, constr
 ## SDK Requirements
 Dart standalone target is supported, but due to some problems with pubspec, the `dart` command must be from Flutter SDK and not Dart SDK. See [dart-lang/pub#3563](https://github.com/dart-lang/pub/issues/3563).
 
+Along with JDK, maven (`mvn` command) is required. On windows, it can be installed using a package manager such as `scoop`.
```
Comment (dcharkes) [inline jnigen/README.md:17]: scoop -> Chocolatey.

File: jnigen/example/in_app_java/src/android_utils/dartjni.h [hunk 6]
```diff
@@ -2,8 +2,9 @@
 // for details. All rights reserved. Use of this source code is governed by a
 // BSD-style license that can be found in the LICENSE file.
 
-#pragma once
+#pragma once
```
Comment (dcharkes) [inline jnigen/example/in_app_java/src/android_utils/dartjni.h:5]: Spurious whitespace at end of line.

(I guess we don't run a formatter for our C files on the CI that complains if files are not formatted correctly. We only do it for Dart files.)

File: jnigen/example/notification_plugin/src/dartjni.h [hunk 7]
```diff
@@ -2,8 +2,9 @@
 // for details. All rights reserved. Use of this source code is governed by a
 // BSD-style license that can be found in the LICENSE file.
 
-#pragma once
+#pragma once
```
Comment (dcharkes) [inline jnigen/example/notification_plugin/src/dartjni.h:5]: whitespace

File: jnigen/example/notification_plugin/src/dartjni.h [hunk 8]
```diff
@@ -2,8 +2,9 @@
 // for details. All rights reserved. Use of this source code is governed by a
 // BSD-style license that can be found in the LICENSE file.
 
-#pragma once
+#pragma once 
 
+// Note: include appropriate system jni.h as found by CMake, not third_party/jni.h
```
Comment (dcharkes) [inline jnigen/example/notification_plugin/src/dartjni.h:7]: Period at end of sentence.

File: jnigen/example/in_app_java/src/android_utils/dartjni.h [hunk 9]
```diff
@@ -2,8 +2,9 @@
 // for details. All rights reserved. Use of this source code is governed by a
 // BSD-style license that can be found in the LICENSE file.
 
-#pragma once
+#pragma once 
 
+// Note: include appropriate system jni.h as found by CMake, not third_party/jni.h
```
Comment (dcharkes) [inline jnigen/example/in_app_java/src/android_utils/dartjni.h:7]: Period.

File: jnigen/example/pdfbox_plugin/.gitignore [hunk 10]
```diff
@@ -35,3 +35,8 @@ build/
 ## Sometimes generated by maven.
 ## TODO(#34) This is unwanted.
 /target/
+
+# Builds & downloads
+*.pdf
+*.jar
+*.exe
```
Comment (dcharkes) [inline jnigen/example/pdfbox_plugin/.gitignore:42]: Newline at end of file.

File: jnigen/example/pdfbox_plugin/jnigen_full.yaml [hunk 11]
```diff
@@ -22,41 +21,17 @@ preamble: |
   // See the License for the specific language governing permissions and
   // limitations under the License.
 
-## Root for generated C bindings.
```
Comment (dcharkes) [inline jnigen/example/pdfbox_plugin/jnigen_full.yaml:25]: Are these comments not useful?

File: jnigen/example/pdfbox_plugin/src/third_party/dartjni.h [hunk 12]
```diff
@@ -2,8 +2,9 @@
 // for details. All rights reserved. Use of this source code is governed by a
 // BSD-style license that can be found in the LICENSE file.
 
-#pragma once
+#pragma once
```
Comment (dcharkes) [inline jnigen/example/pdfbox_plugin/src/third_party/dartjni.h:5]: Same comments as on other copies.

File: jnigen/lib/src/generate_bindings.dart [hunk 13]
```diff
@@ -26,6 +26,7 @@ Future<void> generateJniBindings(Config config) async {
     backend: config.summarizerOptions?.backend,
   );
 
+  // additional sources added using maven downloads and gradle trickery.
```
Comment (dcharkes) [inline jnigen/lib/src/generate_bindings.dart:29]: additional -> Additional

File: jnigen/test/README.md [hunk 14]
```diff
@@ -0,0 +1,11 @@
+Some notes about tests in this directory:
+
+* `jackson_core_test` is an end-to-end test which generates bindings for `jackson_core` library using the whole jnigen pipeline and compares the generated bindings with expected bindings. `simple_package_test` is similar but instead of using a real Java library from maven, uses a stub java class.
```
Comment (dcharkes) [inline jnigen/test/README.md:3]: These could be markdown links so that they are clickable when browsing on Github.

**Review by dcharkes — COMMENTED** [review]
First round (scrolling through the code).

I'll do another round actually playing around with the code.

File: .github/workflows/test-package.yml [hunk 15]
```diff
@@ -192,7 +192,9 @@ jobs:
       - run: Add-Content $env:GITHUB_PATH "$env:JAVA_HOME\bin\server"
       - run: dart pub get
       - run: dart run jni:setup
-      - run: dart test
+      ## -j 1 is required to avoid a rare race condition which is caused by all tests which spawn
```
[the remark this answers is not in the document]
Comment (mahesh-hegde) [reply .github/workflows/test-package.yml:195]: It's #51. My bad, forgot to mention.

File: jnigen/example/pdfbox_plugin/jnigen_full.yaml (hunk 11 above)
Comment (mahesh-hegde) [reply jnigen/example/pdfbox_plugin/jnigen_full.yaml:25]: It refers to jnigen.yaml, which has the explanation. This file is only for generating full package bindings instead of 3 classes. (Maybe this can be just a different command using --override or -D, and better to delete the file).

**Comment by mahesh-hegde** [comment]
Hi, thanks for the review. I will add clang-format checks and update by tomorrow. Feel free to leave more reviews concurrently.

File: README.md (hunk 1 above)
Comment (mahesh-hegde) [reply README.md:17]: I will mention both, because scoop is much cleaner these days.

File: jni/src/README.md (hunk 4 above)
Comment (mahesh-hegde) [reply jni/src/README.md:5]: Will make a small PR later.

File: .github/workflows/test-package.yml [hunk 16]
```diff
@@ -121,11 +121,18 @@ jobs:
         with:
           distribution: 'zulu'
           java-version: '11'
+      - name: install clang tools
+        run: |
+          sudo apt-get update -y
+          sudo apt-get install -y clang-format
```
Comment (dcharkes) [inline .github/workflows/test-package.yml:127]: nice!

File: .github/workflows/test-package.yml [hunk 17]
```diff
@@ -362,7 +371,7 @@ jobs:
           diff -qr _c src/
           diff -qr _dart lib/third_party
       - name: Generate full bindings
-        run: dart run jnigen --config jnigen_full.yaml
+        run: dart run jnigen --config jnigen.yaml --override classes="org.apache.pdfbox.pdmodel;org.apache.pdfbox.text"
```
Comment (dcharkes) [inline .github/workflows/test-package.yml:374]: Why is this needed?

File: README.md [hunk 18]
```diff
@@ -9,16 +9,19 @@ This project intends to provide 2 packages to enable JNI interop from Dart & Flu
 | [jni](jni/) | Ergonomic C bindings to JNI C API and several helper methods |
 | [jnigen](jnigen/) | Tool to generate Dart bindings to Java code using FFI |
 
-This is a work-in-progress project under Google Summer of Code 2022 program.
+This project is a work-in-progress and highly experimental. Users can check out and experiment with the examples. We do not guarantee stability of the API or commands at this point.
 
 ## SDK Requirements
 Dart standalone target is supported, but due to some problems with pubspec, the `dart` command must be from Flutter SDK and not Dart SDK. See [dart-lang/pub#3563](https://github.com/dart-lang/pub/issues/3563).
 
+Along with JDK, maven (`mvn` command) is required. On Windows, it can be installed using a package manager such as `chocolatey` or `scoop`.
```
Comment (dcharkes) [inline README.md:17]: We could actually link to the package

* https://community.chocolatey.org/packages/maven
* https://scoop.sh/#/apps?q=maven

File: jni/lib/src/accessors.dart [hunk 19]
```diff
@@ -0,0 +1,131 @@
...
+void _check(JThrowable exception) {
+  if (exception != nullptr) {
+    env.throwException(exception);
+  }
+}
+
+extension JniResultMethods on JniResult {
+  void check() => _check(exception);
+
+  int get byte {
```
Comment (dcharkes) [inline jni/lib/src/accessors.dart:25]: Nice! Neat way to throw exceptions.

Also a decent way to get "methods" on generated ffi structs. :)

File: jni/src/.clang-format [hunk 20]
```diff
@@ -0,0 +1,15 @@
+# From dart SDK: https://github.com/dart-lang/sdk/blob/main/.clang-format
```
Comment (dcharkes) [inline jni/src/.clang-format:1]: Haha, you make my eyes happy, adopting the formatting from the Dart repo! 👌

File: jnigen/test/README.md [hunk 21]
```diff
@@ -0,0 +1,11 @@
+Some notes about tests in this directory:
+
+* [jackson_core_test](jackson_core_test/) is an end-to-end test which generates bindings for `jackson_core` library using the whole jnigen pipeline and compares the generated bindings with expected bindings. [simple_package_test](simple_package_test/) is similar but instead of using a real Java library from maven, uses a stub java class.
+* [bindings_test.dart](bindings_test.dart) actually runs the generated bindings to make sure they work. There are only a few tests here right now.
```
Comment (dcharkes) [inline jnigen/test/README.md:4]: -= actually

File: jnigen/test/README.md [hunk 22]
```diff
@@ -0,0 +1,11 @@
+Some notes about tests in this directory:
+
+* [jackson_core_test](jackson_core_test/) is an end-to-end test which generates bindings for `jackson_core` library using the whole jnigen pipeline and compares the generated bindings with expected bindings. [simple_package_test](simple_package_test/) is similar but instead of using a real Java library from maven, uses a stub java class.
```
Comment (dcharkes) [inline jnigen/test/README.md:3]: -= real

(the sentence means the same without 'real')

File: jnigen/test/README.md [hunk 23]
```diff
@@ -0,0 +1,11 @@
...
+* [bindings_test.dart](bindings_test.dart) actually runs the generated bindings to make sure they work. There are only a few tests here right now.
+* [test_util/](test_util/) directory contains some code common to both `jackson_core_test` and `simple_package_test`. (Especially the functions `generateAndCompareBindings` and `generateAndAnalyzeBindings`).
+* [yaml_config_test.dart](yaml_config_test.dart) runs the same `jackson_core` configuration but through YAML and makes sure it generates the same bindings.
+* Other files contain unit tests for some error-prone components.
```
Comment (dcharkes) [inline jnigen/test/README.md:7]: Other -> The other

**Review by dcharkes — APPROVED** [review]
lgtm

File: .github/workflows/test-package.yml (hunk 17 above)
Comment (mahesh-hegde) [reply .github/workflows/test-package.yml:374]: I am committing only 2-3 required classes to reduce code.

But in CI we have freedom to generate bindings for the entire `pdmodel` and `text` subpackages and ensure that renaming etc.. are scaling well.

**Comment by mahesh-hegde** [comment]
Done.

**Comment by dcharkes** [comment]
Thanks @mahesh-hegde! 🥁
